File: hub/src/light/effect/translation_effects/constant_spin.py
import math

from light.effect.partial_effect import PartialEffect
from light.led_selector import LEDSelector
from light.effect.resolved_effect import ResolvedEffect
from midi.hit_type import HitType
from light.effect.effect_priority import EffectPriority
from util.scale_functions import ScaleFunctions

class ConstantSpin(PartialEffect):
    def __init__(self, targets, period, speed, num, offset, amplitude=1.0):
        super().__init__(0)
        self.targets = targets
        self.period = period
        self.num = num
        self.offset = offset
        self.amplitude = amplitude

    def getEffect(self, t):
        tNorm = ScaleFunctions.linear(t, self.startTime, self.period)
        ledPositions = []

        for target in self.targets:
            positions = LEDSelector().getAllTargetPositions(target)
            startPos = positions[0]
            ledsPerChild = len(positions) / self.num
            for child in range(0, self.num):
                childPt = startPos + round(child * ledsPerChild + tNorm * len(positions)) + self.offset
                ledPositions.append(LEDSelector().unalias(target, childPt))

        return ResolvedEffect(None, None, list(set(ledPositions)))

    def isTemporal(self):
        return True

    def isComplete(self, t):
        return t > self.startTime + self.period

File: hub/drumlite-js/src/light/effect/positional/ConstantSpin.ts
import { EffectTarget } from "../EffectTarget";
import PartialEffect, { EffectParameters, EffectParameter, defaultMillisecondRange } from "../PartialEffect";
import ResolvedEffect from "../../../effect/ResolvedEffect";
import ScaleFunctions from "../../../util/ScaleFunctions";
import Util from "../../../util/Util";
import LEDSelector from "../../LEDSelector";

export class ConstantSpinParams extends EffectParameters {
    effectName = "Constant Spin";
    typeName = "Positional";

    constructor(targets: EffectTarget[] = [], period: number = 1, num: number = 1, speed: number = 1, offset: number = 0, amplitude: number = 1) {
        super(0);
        this.params.targets = new EffectParameter<EffectTarget[]>("Targets", targets, {type: "target", isArray: true});
        this.params.period = new EffectParameter<number>("Period", period, {range: defaultMillisecondRange});
        this.params.num = new EffectParameter<number>("Number", num);
        this.params.speed = new EffectParameter<number>("Speed", speed);
        this.params.offset = new EffectParameter<number>("Offset", offset);
        this.params.amplitude = new EffectParameter<number>("Amplitude", amplitude);
    }
}

export default class ConstantSpin extends PartialEffect<ConstantSpinParams> {
    constructor(params: ConstantSpinParams, dt = 0) {
        super(params, dt);
    }

    public getEffect(t: number) {
        const {
            targets,
            period,
            num,
            offset,
            amplitude,
            startTime
        } = this.params.params;

        const tNorm = ScaleFunctions.linear(t, startTime.val, period.val);
        const ledPositions: number[] = [];
        const ledSelector = new LEDSelector();

        for (let target of targets.val) {
            const positions = ledSelector.getAllTargetPositions(target);
            const startPos = positions[0];
            const ledsPerChild = positions.length / (num.val);
            for (let child of Util.range(0, num.val)) {
                const childPt = startPos + Math.round(child * ledsPerChild + tNorm * positions.length) + (offset.val);
                ledPositions.push(ledSelector.unalias(target, childPt));
            }
        }

        return [[new ResolvedEffect(undefined, undefined, [...new Set<number>(ledPositions)])]];
    }

    public isTemporal() {
        return true;
    }

    public isComplete(t: number) {
        return t > this.params.params.startTime.val + (this.params.params.period.val);
    }
}
